### PR TITLE
Move a number of JsonPropertyInfo reflection components to DefaultJsonTypeInfoResolver

### PR DESCRIPTION
**Pull request by eiriktsarpalis**

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs [hunk 1]
```diff
@@ -81,8 +82,7 @@ protected override void InitializeConstructorArgumentCaches(ref ReadStack state,
         {
             JsonTypeInfo typeInfo = state.Current.JsonTypeInfo;
 
-            typeInfo.CreateObjectWithArgs ??=
-                JsonSerializerOptions.MemberAccessorStrategy.CreateParameterizedConstructor<T, TArg0, TArg1, TArg2, TArg3>(ConstructorInfo!);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs:85]: This was calling into reflection in the deserialization hot path -- replaced with an override to the JsonTypeInfo reflection configuration callback.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonMetadataServices.Helpers.cs [hunk 2]
```diff
@@ -166,5 +166,63 @@ private static void PopulateProperties(JsonTypeInfo typeInfo, Func<JsonSerialize
 
             // NB we don't need to sort source gen properties here since they were already sorted at compile time.
         }
+
+        private static JsonPropertyInfo<T> CreatePropertyInfoCore<T>(JsonPropertyInfoValues<T> propertyInfoValues, JsonSerializerOptions options)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonMetadataServices.Helpers.cs:170]: Replaces a `JsonPropertyInfo<T>` constructor that was specifically used for populating sourcegen metadata.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs:85]: > This was calling into reflection in the deserialization hot path

For source-gen I believe the property would not have been `null`. Is the reflection now trimmed out/any size savings?

**Review by layomia — APPROVED** [review]
LGTM.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs:85]: > For source-gen I believe the property would not have been null

You're right, it appears we're only using the large constructor converter in source gen.

> Is the reflection now trimmed out/any size savings?

I haven't run any tests, but I figure it should now able to trim all the Ref.Emit components (warnings on that particular `MemberAccessorStrategy` property were being suppressed for some reason). 

cc @eerhardt

**Pull request merged**